NEW Subtotal module

YannisHoareau opened the pull request:
This PR is to add a new module named subtotal (in compliance with #30742). This covers proposals, orders, shipments, invoices, and rec_invoices. This module allows you to add titles and subtotals lines to structure involved documents. It is also possible to add options to the lines to structure the linked PDF.

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Changes to follow mvc logic

YannisHoareau committed:
Reworked admin page and form to add lines

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Reworked adding line logic

YannisHoareau committed:
Adding options when editing subtotal lines

YannisHoareau committed:
Fix translations

YannisHoareau committed:
Fixed errors/displays and started pdf

YannisHoareau committed:
Color for subtotals pdf lines

YannisHoareau committed:
FIX display of subtotal totht

YannisHoareau committed:
Added pdf azur for propal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix duplicate translation

YannisHoareau committed:
Added subtotal support for facture pdf

YannisHoareau committed:
Added subtotal support for commande pdf

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Improve UI/translations

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Restored old pdf

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Info to warn user for unsupported pdf

YannisHoareau committed:
Added title lines VAT rate and discount support

This is meant for future feature wich is block mass changing vat rate
and discount percentage

YannisHoareau committed:
Working on block apply vat and discount

YannisHoareau committed:
Added buttons for block actions

YannisHoareau committed:
Handle editing vat and discount for subtotal lines

YannisHoareau committed:
Editing vat and discount for subtotal lines bloc working

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Added possibility to move by block

YannisHoareau committed:
Updated adding and updating a subtotal line

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Improved vat and discount block update

YannisHoareau committed:
Improvement for block vat/discount and line edition

Editing a title line edits the corresponding subtotal line

YannisHoareau committed:
Improvement for moving by block

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Bad tile or st line placement managing

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Improved bad title or st line placement managing

YannisHoareau committed:
Adding subtotal line improved

Adding a subtotal line adds it right under its corresponding title

YannisHoareau committed:
Improved deleting subtotal line

Added possibility to choose if you want to delete the corresponding
subtotal line when deleting a title line

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Preventing too high level titles to be created

YannisHoareau committed:
Create and update line errors managing

YannisHoareau committed:
Improved bad title or st line placement managing

YannisHoareau committed:
Improved st line creation

YannisHoareau committed:
improve headers

YannisHoareau committed:
fix bad block placement managing

YannisHoareau committed:
fix adding st line not working if duplicates titles

YannisHoareau committed:
fix translations

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix block update

YannisHoareau committed:
FIX special chars bug

If special char like " ' " was used in title it could be converted to
special char entity.

YannisHoareau committed:
Prepare for pdf options

Pdf options like page break befor title should be shown as a picto if activated on a subtotal line on a document

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'dolibarr-new-subtotal' of github.com:YannisHoareau/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Added option managing

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Improved subtotal options and PDF integration

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Code refactor

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Action name/Error managment/PDF refreshing

When adding or updating a subtotal line

YannisHoareau committed:
Reformat

YannisHoareau committed:
Reworked subtotal options database managment

YannisHoareau committed:
Changed access to special code

YannisHoareau committed:
Remove unecessary call to php trait

YannisHoareau committed:
Changed definition of subtotals special code constant

YannisHoareau committed:
Reworked align on PDF

YannisHoareau committed:
Removed unecessary function and improved error managment

YannisHoareau committed:
Typo fix and removed treated todos

YannisHoareau committed:
Post typo changed for to better match subtotals names

YannisHoareau committed:
Disabled edit if status is not draft

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Changed way subtotals options are stored

YannisHoareau committed:
Added view managing when creating a document form an other

YannisHoareau committed:
Improved creating document from an other

Can check subtotals lines with table head checkbox and removed highlight
class for better UI.

YannisHoareau committed:
Make include of subtotals tpl more clear

YannisHoareau committed:
Manage centered or justified case

If user chose to center or justify, we don't change nothing

YannisHoareau committed:
Improved pdf

YannisHoareau committed:
Removed unecessary code block

YannisHoareau committed:
optimisation

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Code sniffer fix

YannisHoareau committed:
Code sniffer fix

YannisHoareau committed:
Code sniffer fix

YannisHoareau committed:
Code sniffer fix

YannisHoareau committed:
Code sniffer fix and added missing translation

YannisHoareau committed:
Fix php code sniffer

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Reload page when setup saved on subtotals admin page

YannisHoareau committed:
Fix php code sniffer

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
pjan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
phan fix

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
php warning fix

YannisHoareau committed:
php warning fix

YannisHoareau committed:
php phan fix

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
php phan fix

YannisHoareau committed:
php phan fix

YannisHoareau committed:
Fix bug admin page not loading because of const not defined

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
php phan fix

YannisHoareau committed:
php phan fix

YannisHoareau committed:
FIX subtotals admin page display

YannisHoareau committed:
php phan fix

YannisHoareau committed:
FIX php phan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix bugs and langs

YannisHoareau committed:
Fix bug pdf align

YannisHoareau committed:
Replace include by require

YannisHoareau committed:
Add headers to avoid refreshing and adding unwanted lines

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Add GETPOST check for security

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Subtotal option when creating a document from another

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Add field subtotal options for subtotal lines

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Update to follow mvc

YannisHoareau committed:
Fix php phan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix php phan

YannisHoareau committed:
Fix phpstan/phan

YannisHoareau committed:
Fix phpstan

YannisHoareau committed:
Fix phpstan

YannisHoareau committed:
Fix phpstan

YannisHoareau committed:
Fix phpstan

YannisHoareau committed:
Fix phpstan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Update to switch to extraparams

YannisHoareau committed:
Update to switch to extraparams

YannisHoareau committed:
Update to switch to extraparams

YannisHoareau committed:
Retrieve extraparams from db to objectline

YannisHoareau committed:
Modified last things to switch to extraparams

YannisHoareau committed:
Cleaning unnecessary code lines

YannisHoareau committed:
Fix php-stan

YannisHoareau committed:
Section subtotal in extraparams to differentiate if needed for further devs

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Keep extraparams when creating from another object

YannisHoareau committed:
Change default value to false when creating a subtotal line

YannisHoareau committed:
Fix clone would not keep extraparams in new object

YannisHoareau committed:
Fix dark subtotal line background color

Fix when a subtotal background color is too dark and edit pencil or
delete trash could not be seen

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix typo

YannisHoareau committed:
Fix typo

YannisHoareau committed:
Added subtotals for facturerec

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'dolibarr-new-subtotal' of github.com:YannisHoareau/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix precommit

YannisHoareau committed:
Added extraparams when creating rec from fac and other way

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix objectline null

YannisHoareau committed:
Desactivating block vat / discount update for facturerec

YannisHoareau committed:
reformating code

YannisHoareau committed:
Added expeditions for subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Save extraparams for shipping lines

YannisHoareau committed:
Display of subtotals lines in shipments

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Display when creating facture from shipments

YannisHoareau committed:
Improve display of lines and invoice creation from shipments

YannisHoareau committed:
Fix error if missing line rang

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Deleted duplicate

YannisHoareau committed:
Added deletion of subtotal line in shipping documents

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Not including subtotal lines if there is no product line in between

YannisHoareau committed:
Update get subtotal lines in shipment docs to disable

YannisHoareau committed:
Delete possibility to edit subtotal lines in shipments

YannisHoareau committed:
Handle pdf for shipment

YannisHoareau committed:
Handle conf stock or shipment supporting services

YannisHoareau committed:
Fix precommit

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix duplicate name creating bug

YannisHoareau committed:
Fix bug where id could be changed by the line id and would create bug

YannisHoareau committed:
Deleted subtotal lines when STOCK_SUPPORTS_SERVICES is enabled and block would only have service lines

YannisHoareau committed:
Disable shipments in admin modules

Shimpements subtotals lines are only created from commands

YannisHoareau committed:
Fixing phan and stan

YannisHoareau committed:
Deleted unused template and phan fix

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix phan

frederic34 commented:
![image](https://github.com/user-attachments/assets/cdd6ddc3-995e-4a3d-9bc6-2115c389982e)
https://github.com/Dolibarr/dolibarr/pull/33504

YannisHoareau commented:
Thank you @frederic34, I had planned to do this fix but will update the branch when your PR is accepted.

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

marc-dll committed:
FIX: phan

marc-dll committed:
Merge branch 'develop' of https://github.com/YannisHoareau/dolibarr into dolibarr-new-subtotal

marc-dll committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into dolibarr-new-subtotal

marc-dll committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix template bug

If document was not in subtotal scope and would use a template used by
subtotal it would create an error.

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau commented:
Hello @eldy, all the tests are now passing except if the error come from the develop branch, so now that it is almost stable, can you provide some feedbacks about this PR please ?

frederic34 commented:
@YannisHoareau 
![image](https://github.com/user-attachments/assets/575b083c-a390-4cb6-9653-b9db47f1425e)
with margin module
![image](https://github.com/user-attachments/assets/163f4b87-b6b9-44e7-8798-c77fdd8ac48e)

eldy committed:
Merge branch 'develop' into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix php-stan

YannisHoareau committed:
Fix bad display when modules and confs were activated

YannisHoareau committed:
Fix php codesniffer

YannisHoareau commented:
@frederic34 Display errors should be fixed

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

eldy commented:
Hi @YannisHoareau 
Another PR to all mass update of mark rate is in conflict with yours. Can you help me to solve the conflict ?

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Excluding subtotal lines when mass updating

eldy commented on htdocs/expedition/card.php:
```diff
@@ -2440,7 +2512,7 @@
 		}
 
-		if (empty($reshook)) {
+		if (empty($reshook) && empty($lines[$i]->product_type)) {
```
I don't understand this addition in test 
`&& empty($lines[$i]->product_type)`
What if product_type is 1 (for a service) ?
May be you want to do 
`&& $lines[$i]->product_type != 9 ?`

YannisHoareau replied:
Yes you are right, I am fixing this.

YannisHoareau committed:
Fix shipments service lines exluded + showing subtotal line with specific configs

hregis commented:
@grandoc c'est la mort de Milestone dont subtotal était un clone... À quand la mort de Multicompany ?

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

eldy commented on htdocs/comm/propal/class/propal.class.php:
```diff
@@ -841,6 +869,13 @@ public function addline($desc, $pu_ht, $qty, $txtva, $txlocaltax1 = 0.0, $txloca
 					}
 
 					$this->lines[] = $this->line;
+				} else {
```
I don't understand the goal of this portion of added code ?
Once we have inserted line (code line 857), it it is not in a clone process, we must recalculate the position (updateRangOfLine).
If not, whe should have nothing to do. The line has been inserted by the line->insert with all similar value than the cloned object.

YannisHoareau commented:
Hey @eldy, do you have any news about this PR ? There is no conflict or CI error anymore, is there anything to solve ?

thomas-Ngr commented:
@YannisHoareau Hi, thanks for this feature !

I made some tests (only on an invoice). This is the invoice I created : 

![subtotal_lignes](https://github.com/user-attachments/assets/2c9ce31e-b648-4d42-898d-2598a2c8919b)


- the display of subtotals seems not to work in ODTs (no way to know it is a subtotal, no display of the total value)
![subtotal_odt](https://github.com/user-attachments/assets/0509a586-fb14-46e9-9b55-d74574dcbfad)

- the display of subtotals and of lines inbetween seems not to work in PDFs
![subtotal_pdf](https://github.com/user-attachments/assets/296a03bf-3d48-45f5-82ed-642f3683c12a)

- there is a strange behavior just after editing a line : only subtotal lines are displayed. One needs to reload the page to display product lines.
![subtotal_modify_line](https://github.com/user-attachments/assets/289f4eab-a74d-4b85-b29c-76b3a6c3085b)

- It seems one should be able to change VAT rate or discount rate per chapter. However, when I try to do it, I get an error : 
![subtotal_change_chapter_vat](https://github.com/user-attachments/assets/af29a19a-17d6-4cb3-a4bd-e6c500b187e1)

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix line display with situation invoices + bug block line update

YannisHoareau committed:
Added ODT managment

This works with invoices but has not been tested with other documents

YannisHoareau committed:
Fix bug when editing VAT/discount by block

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix error raised when subtotal line added

Subtotal line had no fk_product and raised the error but we want this
line to have no fk product

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
Fix precommit

YannisHoareau committed:
Added ODT template for documents that uses subtotals

This template can be usefull to understand how to create an odt template
using subtotals module

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Fix phan

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
fix: buttons showing in bad document status

YannisHoareau committed:
clean: unwanted commited files

YannisHoareau committed:
add: table examples to use with subtotals on ODT templates

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
fix: Unwanted print of value when creating a document from another

YannisHoareau committed:
feat: renaming for better understanding

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau committed:
feat: Improved templates for documents related to subtotals

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

YannisHoareau commented:
Hello @eldy, do you have any news about this PR ?

YannisHoareau committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into dolibarr-new-subtotal

eldy committed:
Merge branch 'develop' into dolibarr-new-subtotal

eldy committed:
Merge branch 'develop' into dolibarr-new-subtotal
